homeassistant/components/omnilogic/sensor.py

```diff
@@ -0,0 +1,489 @@
+"""Definition and setup of the Omnilogic Sensors for Home Assistant."""
+
+import logging
+
+from homeassistant.components.sensor import DEVICE_CLASS_TEMPERATURE
+from homeassistant.const import (
+    CONCENTRATION_PARTS_PER_MILLION,
+    MASS_GRAMS,
+    PERCENTAGE,
+    TEMP_CELSIUS,
+    TEMP_FAHRENHEIT,
+    VOLUME_LITERS,
+)
+
+from .common import OmniLogicEntity, OmniLogicUpdateCoordinator
+from .const import COORDINATOR, DOMAIN, PUMP_TYPES
+
+TEMP_UNITS = [TEMP_CELSIUS, TEMP_FAHRENHEIT]
+PERCENT_UNITS = [PERCENTAGE, PERCENTAGE]
+SALT_UNITS = [f"{MASS_GRAMS}/{VOLUME_LITERS}", CONCENTRATION_PARTS_PER_MILLION]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the sensor platform."""
+
+    coordinator = hass.data[DOMAIN][entry.entry_id][COORDINATOR]
+    sensors = []
+
+    for this_entity in coordinator.data:
+        entity_data = coordinator.data[this_entity]
+
+        if entity_data["type"] == "backyard":
```

Comment:
Make a dict that maps the length of the global id and sub id name to entity classes and settings.

```py
SENSOR_TYPES = {
	(2, "Backyard"): {
		"entity_classes": {"airTemp": OmniLogicTemperatureSensor},
		"name": "Air Temperature",
		"device_class": DEVICE_CLASS_TEMPERATURE,
		"icon": None,
		"unit": TEMP_FAHRENHEIT,
		"guard_condition": {},
	},
	(4, "BOWS"): {
		"entity_classes": {"waterTemp": OmniLogicTemperatureSensor},
		"name": "Water Temperature",
		"device_class": DEVICE_CLASS_TEMPERATURE,
		"icon": None,
		"unit": TEMP_FAHRENHEIT,
		"guard_condition": {},
	(6, "Chlorinator"): {
		"entity_classes": {"Timed-Percent": OmniLogicChlorinatorSensor},
		"name": "Setting",
		"device_class": None,
		"icon": "mdi:gauge",
		"unit": PERCENTAGE,
		"guard_condition": {"Shared-Type": "BOW_SHARED_EQUIPMENT", "status": "0"},
	...
}

entities = []

for item_id, item in coordinator.data.items():
    id_len = len(item_id)
    item_kind = item_id[-2]
    entity_settings = SENSOR_TYPES.get((id_len, item_kind))
    if not entity_settings:
        continue
	for state_key, entity_class in entity_settings["entity_classes"].items():
		if state_key not in item:
			continue

		# On some conditions we ignore the item.
		guard_condition = entity_settings["guard_condition"]
		if all(
			item.get(guard_key) == guard_value
			for guard_key, guard_value in guard_condition.items()
		):
			continue
			
		entity = entity_class(
			coordinator,
            state_key,
			entity_settings["name"],
			item_id
			entity_settings["device_class"],
			entity_settings["icon"],
            entity_settings["unit"],
		)
		entities.append(entity)

```

Reply:
Trying to follow your suggestions here - where are you getting the number in the SENSOR_TYPES dict? I understand it is the length of the global ID, but I am struggling to understand how to know what length will align to what sensor type?

If I'm following correctly it is based on the depth or order of where we will find each item, but I don't have confidence that the msp_system_id (backyard system_id) will be a consistent length from Hayward nor that the data returned from Hayward will always follow a consistent order or pattern. Maybe I'm missing something.

Reply:
The id keys will look something like this (incomplete):
```
[('Backyard', '24353'),
 ('Backyard', '24353', 'BOWS', '1'),
 ('Backyard', '24353', 'BOWS', '1', 'Filter', '2'),
 ('Backyard', '24353', 'BOWS', '1', 'VirtualHeater', '15'),
 ('Backyard', '24353', 'BOWS', '1', 'Heater', '16'),
 ('Backyard', '24353', 'BOWS', '1', 'CSAD', '0'),
 ('Backyard', '24353', 'BOWS', '1', 'Group', '34'),
 ('Backyard', '24353', 'BOWS', '1', 'Lights', '11'),
 ('Backyard', '24353', 'BOWS', '1', 'Lights', '25'),
 ('Backyard', '24353', 'BOWS', '1', 'Relays', '5'),
 ('Backyard', '24353', 'BOWS', '1', 'Relays', '14'),
 ('Backyard', '24353', 'BOWS', '1', 'Relays', '26'),
 ('Backyard', '24353', 'Relays', '17'),
 ('Backyard', '36211'),
 ('Backyard', '36211', 'BOWS', '2'),
 ('Backyard', '36211', 'BOWS', '2', 'Filter', '3'),
 ('Backyard', '36211', 'BOWS', '2', 'VirtualHeater', '4'),
 ('Backyard', '36211', 'BOWS', '2', 'Heater', '5'),
 ('Backyard', '36211', 'BOWS', '2', 'CSAD', '0'),
 ('Backyard', '36211', 'BOWS', '2', 'Lights', '23'),
 ('Backyard', '36211', 'BOWS', '2', 'Lights', '24'),
 ('Backyard', '36211', 'BOWS', '2', 'Relays', '7'),
 ('Backyard', '36211', 'Relays', '14'),
 ('Backyard', '36211', 'Relays', '29')]
```
We can deduce this structure by looking at the input data before we parse it. Each item in the data tree that has a system id marks a level in the id tuple.

Reply:
So if i extrapolate your dict just to make sure I'm following, the next items would be:

```
SENSOR_TYPES = {
	(2, "Backyard"): {
		"entity_classes": {"airTemp": OmniLogicTemperatureSensor},
		"name": "Air Temperature",
		"device_class": DEVICE_CLASS_TEMPERATURE,
		"icon": None,
		"unit": TEMP_FAHRENHEIT,
		"guard_condition": {},
	},
	(4, "BOWS"): {
		"entity_classes": {"waterTemp": OmniLogicTemperatureSensor},
		"name": "Water Temperature",
		"device_class": DEVICE_CLASS_TEMPERATURE,
		"icon": None,
		"unit": TEMP_FAHRENHEIT,
		"guard_condition": {},
	},
	(6, "Chlorinator"): {
		"entity_classes": {"Timed-Percent": OmniLogicChlorinatorSensor},
		"name": "Setting",
		"device_class": None,
		"icon": "mdi:gauge",
		"unit": PERCENTAGE,
		"guard_condition": {"Shared-Type": "BOW_SHARED_EQUIPMENT", "status": "0"},
	},
       (6, "Filter"): {
		"entity_classes": {"Timed-Percent": OmniLogicPumpSpeedSensor},
		"name": "Speed",
		"device_class": None,
		"icon": "mdi:speedometer",
		"unit": PERCENTAGE,
		"guard_condition": {},
	},
	...
}
```

Reply:
And also for clarity there would be two items in dict for relays, one as (4, "Relays"): {... and (6, "Relays"): {... to capture the different places they might appear?

Reply:
The `state_key` for the filter pump speed sensor should be `"filterSpeed"`. Otherwise looks good.

Reply:
One other issue I see here (or don't understand) - for some item types (Chlorinator, CSAD) we have multiple sensors within them (Chlorinator Setting, Chlorinator Salt Level / CSAD pH and CSAD ORP)... within this dict structure would this then mean that I would have:

```
(6, "Chlorinator"): [
  {
    "entity_classes": {"Timed-Percent": OmniLogicChlorinatorSensor},
    "name": "Setting",
    "device_class": None,
    "icon": "mdi:gauge",
    "unit": PERCENTAGE,
    "guard_condition": {"Shared-Type": "BOW_SHARED_EQUIPMENT", "status": "0"},
  },
  {
    "entity_classes": {"avgSaltLevel": OmniLogicSaltLevelSensor},
    "name": "Salt Level",
    "device_class": None,
    "icon": "mdi:gauge",
    "unit": NONE,
    "guard_condition": {"Shared-Type": "BOW_SHARED_EQUIPMENT", "status": "0"},
  },
]
```

Reply:
I only added one entity class item for the Chlorinator for brevity, but it should be two with each different state keys.

Reply:
Right, we need different setting for each kind so we need to do it like in your example, above.

Reply:
That's where I'm confused... so there would be two `(6, "Chlorinator"):` items in the dict? Or would I have to adapt the data update coordinator to issue multiple state keys for the Chlorinator entity (in which case I see this starting to get complicated once we add more platforms).

Reply:
Confirming:

I should add the multiple sensors as a list in the dict as I highlighted and then modify the method to handle both sensors when there is a list vs a single item correct?

Reply:
Make it always a list of dict settings.

Reply:
I have implemented this approach in the latest commit.